mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -738,11 +741,11 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
   if (!tensorType)
     return {};
 
-  // Out of bound indices produce undefined behavior but are still valid IR.
-  // Don't choke on them.
+  // Fold dim to posion if the index is out of bound. Poison represents
+  // undefined behavior.
   int64_t indexVal = index.getInt();
   if (indexVal < 0 || indexVal >= tensorType.getRank())
-    return {};
+    return ub::PoisonAttr::get(getContext());
```

Comment:
Is the Tensor dialect already dependent on UB dialect and generating poison values?

Reply:
This is really a design question here: this new addition can be a problem for some user, it's not clear to me if we can actually use poison everywhere in MLIR right now.

Reply:
I read the code comment again and think you are right.  It seems `tensor.dim` ops with out-of-bound indices are treated as valid IR and may have following handling solutions:
- Compiled to runnable binary and raise exception on runtime
- Handled in other passes to raise warning or errors on compile time depends on the compiler design.

So should I just simply fix the out-of-bound access bug in the PatternRewriter `FoldDimOfAllocTensorOp` and keep this folding behavior as before. What's your opinion?

Reply:
Another question: is `ub.poison` is designed to be a valid IR and able to be compiled to a binary?

Reply:
After reviewing the design related to Poison semantics, there is no clear definition of the issue of excessive crossing. Can we simply skip folding and do nothing according to the original logic, so as not to disrupt the processing of the original code.

Reply:
>  It seems tensor.dim ops with out-of-bound indices are treated as valid IR 

It is documented as undefined behavior, so turning it into poison is clearly allowed by the spec.
However at the moment we haven't deployed the UB dialect as a really "central" component, and "fold" is, even more than canonicalization, very anchored in the core of MLIR (it's not optional, can't be disabled, etc.). 
So folding to poison makes me a bit nervous here, if there were already other folding of the tensor dialect to poison that wouldn't be a problem.

> Can we simply skip folding and do nothing 

This is the most conservative solution IMO.